### PR TITLE
Mssw chapter05

### PR DESCRIPTION
**Pull request by Spinodal**

### CODE REVIEW

File: chapter05/mssw/exercise2.pl [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+    # check if the line starts with a '#' and skip it
+    if ( $row =~ /^\s*#/ ) {
+        next;
+    }
+    my ( $source, $destination, $bytes ) = split /\s+/, $row;
+    $total_bytes{$source}{$destination} += $bytes;
+}
+
+# Lets create a complex hash! \o/
```
Comment (ardavey) [inline chapter05/mssw/exercise2.pl:26]: Bonus points if you draw a PeGS diagram for this...

File: chapter05/mssw/exercise2.pl [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    print ">> Total for source '$source' is: $report{$source}->{total}\n";
+    foreach my $destination (
+        sort {
+            $report{$source}->{destinations}{$b}
+              <=> $report{$source}->{destinations}{$a}
+        }
+        keys %{ $report{$source}->{destinations} }
+      )
+    {
+        printf( "%-30s => %30s : %10d bytes\n",
```
Comment (ardavey) [inline chapter05/mssw/exercise2.pl:61]: Doesn't quite match the sample output - the Professor's parser is going to go coco-nuts!

Comment (Spinodal) [inline chapter05/mssw/exercise2.pl:61]: i think i originally had it like they did, but hte changing line lengths made it really hard to check if everything was working properly :stuck_out_tongue:

File: chapter05/mssw/exercise2.pl (hunk 1 above)
Comment (Spinodal) [inline chapter05/mssw/exercise2.pl:26]:
```
      +---------+                                                                                           
      | %report +-->                                                                                           
      +----+----+                                                                                           
           |                                                                                                
+----------+----------+                                                                                     
+--------+   +--------+                                                                                     
| source +-->|   +---------------+-------------------------------+                                          
+--------+   +--------+          +--------------+   +------------+                                          
|        |   |        |          | destinations +-->|      +-----------+-----------------------------------+
|        |   |        |          +--------------+   +------------+     +------------------+  +-------------+
|        |   |        |          | total        +-->| 120000     |     | destination.host +->+ 1452        |
|        |   |        |          +--------------+   +------------+     +------------------+  +-------------+
|        |   |        |                                                | other.destination+->+ 1005        |
|        |   |        |                                                +------------------+  +-------------+
|        |   |        |                                                                                     
|        |   |        |                                                                                     
|        |   |        |                                                                                     
|        |   |        |                                                                                     
|        |   |        |                                                                                     
+--------+   +--------+                                                                                     
| other.s|-->|   +---------------+-------------------------------+                                          
+--------+   +--------+          +--------------+   +------------+                                          
                                 | destinations +-->|     +------------+-----------------------------------+
                                 +--------------+   +------------+     +-------------------+  +------------+
                                 | total        +-->| 150657     |     | blargstation      +->+ 10546873   |
                                 +--------------+   +------------+     +-------------------+  +------------+

```

File: chapter05/mssw/exercise2.pl (hunk 2 above)
Comment (ardavey) [inline chapter05/mssw/exercise2.pl:61]: I mostly meant the word "bytes" being there :smile:

Comment (Spinodal) [inline chapter05/mssw/exercise2.pl:61]: oooh right, i think i must have copied and pasted that from the page before and then it just stuck

File: chapter05/mssw/exercise2.pl [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+#     total = $total_bytes,
+#   },
+#   ...
+# );
+my %report;
+foreach my $source ( sort keys %total_bytes ) {
+    for my $destination ( sort keys %{ $total_bytes{$source} } ) {
+        $report{$source}->{destinations}{$destination} +=
+          $total_bytes{$source}{$destination};
+        $report{$source}->{total} += $total_bytes{$source}{$destination};
```
Comment (thruddle) [inline chapter05/mssw/exercise2.pl:42]: You don't seem to be able to decide whether or not to include the dereferencing arrows, or to use let them be implicit.
It doesn't really matter which you go for, but you should try to be consistent.

Comment (Spinodal) [inline chapter05/mssw/exercise2.pl:42]: I thought you need at least one arrow to get the party started? seeing as `%report` isn't a reference.... although i see what you mean now, i seem to have missed a couple....

Comment (utopia-2) [inline chapter05/mssw/exercise2.pl:42]: @Spinodal You only need to dereference the first reference in a data structure. As `%report` is a hash, no explicit dereferencing needed :smile:.

File: chapter05/mssw/exercise3.pl [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+          $total_bytes{$source}{$destination};
+        $report{$source}->{total} += $total_bytes{$source}{$destination};
+    }
+}
+
+open( my $write_fh, '>:encoding(UTF-8)', $output )
+  or die "could not create file '$output' $!";
+
+# Now lets redump the file
+foreach my $source ( sort { $report{$b} cmp $report{$a} } keys %report ) {
```
Comment (SimonGross) [inline chapter05/mssw/exercise3.pl:52]: This was supposed to sort based on the name of the source machine, but I think this is sorting it based on the hash ref values of the destination hash, not the key names. 
This should be: `sort { $a cmp $b} keys %report` or just `sort keys %report`

**Pull request merged**